doc/developer/design/20231110_aws_connections.md

```diff
@@ -0,0 +1,175 @@
+# AWS Connection
+
+## The Problem
+
+Github issue:  [Issue 23055](https://github.com/MaterializeInc/materialize/issues/23055)
+
+We plan to build an AWS S3 sink, in which we write to an S3 bucket under the end user's
+control. We'll need to integrate with AWS IAM so that our users can grant their Materialize
+region the ability to write to their desired S3 bucket.
+
+Integration with AWS IAM would also be useful for facilitating IAM authentication with the
+following existing source types:
+
+- PostgreSQL on AWS RDS
+- Kafka on AWS MSK
+
+While we support AWS RDS/MSK today via traditional username/password authentication, many
+users prefer to use AWS IAM with these services, as it eliminates the need to manage
+additional user accounts/passwords.
+
+## Success Criteria
+- Be able to create an AWS connection using either AWS credentials or by using the [AssumeRole API](https://docs.aws.amazon.com/STS/latest/APIReference/API_AssumeRole.html).
+- Be able to check that the AWS connection is valid or not using `VALIDATE CONNECTION`.
+
+## Solution Proposal
+Allow users to create an AWS CONNECTION with AWS Credentials or AWS IAM AssumeRole.
+
+### AWS Credentials
+Creating an AWS connection with credentials should be quite straightforward. The SQL would look like,
+```
+CREATE SECRET aws_secret_key = 'secret';
+
+CREATE CONNECTION aws_conn TO AWS (
+  ACCESS KEY ID = 'access_key',
+  SECRET ACCESS KEY = SECRET aws_secret_key
+);
+
+```
+The `aws_conn` is now available to be used in supported queries.
+
+### AWS IAM AssumeRole
+Users will need to first create a role on their end, say, `MaterializeConn` and note down it's arn, e.g: `arn:aws:iam::001234567890:role/MaterializeConn`.
+
+User then creates a connection in Materialize with that arn.
+```
+CREATE CONNECTION aws_conn TO AWS (
+  ASSUME ROLE ARN = 'arn:aws:iam::001234567890:role/MaterializeConn'
+);
+```
+On Materialize's end we generate an External ID and create an entry in a new `mz_aws_connections` table.
+Users should be able to get the external_id and the principal by querying this table.
+```
+SELECT principal, external_id
+FROM mz_aws_connections aws
+JOIN mz_connections c ON aws.id = c.id
+WHERE c.name = 'aws_conn';
+```
+
+Users should then be able to add a trust policy in their AWS account to give access to Materialize.
+```
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Sid": "",
+      "Effect": "Allow",
+      "Principal": {
+        "AWS": "<principal>"
+      },
+      "Action": "sts:AssumeRole",
+      "Condition": {
+        "StringEquals": {
+          "sts:ExternalId": "<external_id>"
+        }
+      }
+    }
+  ]
+}
+```
```

Comment:
We should explicitly give them this assume role policy in a form that can be easily copy/pasted. We should not expect them to fill in the blanks.

Reply:
Do you mean to say we give them the entire generated xml with the principal/external ID?

Reply:
This is json, but yes.

Reply:
lol yes json 🤦‍♀️ (I have too many aws request xml tabs open).

Reply:
I will see how to do this best while implementing this. I am not very keen on storing the entire json in our system, because we don't own the format and it can change. Maybe in the docs we give them a sql to run with the entire json in it.

Reply:
I added this to the proposal in https://github.com/MaterializeInc/materialize/pull/23413/files#diff-67c46a1362121b3b9802f9fe608bd0cc9876f37c15c0e3559d1302b2c8f54a76R83.